plugin/pkg/admission/initialresources/data_source.go

```diff
@@ -0,0 +1,205 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package initialresources
+
+import (
+	"flag"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	influxdb "github.com/influxdb/influxdb/client"
+	"k8s.io/kubernetes/pkg/api"
+)
+
+const (
+	cpuSeriesName      = "autoscaling.cpu.usage.1m"
+	memSeriesName      = "autoscaling.memory.usage.1m"
+	cpuContinuousQuery = "select derivative(value) as value from \"cpu/usage_ns_cumulative\" where pod_id <> '' group by pod_id, container_name, container_base_image, time(1m) into " + cpuSeriesName
+	memContinuousQuery = "select mean(value) as value from \"memory/usage_bytes_gauge\" where pod_id <> '' group by pod_id, container_name, container_base_image, time(1m) into " + memSeriesName
+	timeFormat         = "2006-01-02 15:04:05"
+)
+
+var (
+	influxdbHost = flag.String("ir-influxdb-host", "localhost:8080/api/v1/proxy/namespaces/kube-system/services/monitoring-influxdb:api", "Address of InfluxDB which contains metrics requred by InitialResources")
+	user         = flag.String("ir-user", "root", "User used for connecting to InfluxDB")
+	// TODO: figure out how to better pass password here
+	password = flag.String("ir-password", "root", "Password used for connecting to InfluxDB")
+	db       = flag.String("ir-dbname", "k8s", "InfluxDB database name which contains metrics requred by InitialResources")
+)
+
+// WARNING: If you are planning to add another implementation of dataSource interface please bear in mind,
+// that dataSource will be moved to Heapster some time in the future and possibly rewritten.
+type dataSource interface {
+	// Returns <perc>th of sample values which represent usage of <kind> for containers running <image>,
+	// withing time range (start, end), number of samples considered and error if occured.
+	// If <exactMatch> then take only samples that concern the same image (both name and take are the same),
+	// otherwise consider also samples with the same image a possibly different tag.
+	GetUsagePercentile(kind api.ResourceName, perc int64, image string, exactMatch bool, start, end time.Time) (usage int64, samples int64, err error)
+}
+
+func newDataSource(kind string) (dataSource, error) {
+	if kind == "influxdb" {
+		return newInfluxdbSource(*influxdbHost, *user, *password, *db)
+	}
+	if kind == "gcm" {
+		return newGcmSource()
+	}
+	return nil, fmt.Errorf("Unknown data source %v", kind)
+}
+
+// TODO(piosz): rewrite this once we will migrate into InfluxDB v0.9.
+type influxdbSource struct {
+	conf *influxdb.ClientConfig
+	conn *influxdb.Client
+	// protects <conn>
+	mutex sync.RWMutex
+}
+
+func newInfluxdbSource(host, user, password, db string) (dataSource, error) {
+	conf := &influxdb.ClientConfig{
+		Host:     host,
+		Username: user,
+		Password: password,
+		Database: db,
+	}
+	source := &influxdbSource{
+		conf:  conf,
+		mutex: sync.RWMutex{},
+	}
+	go source.ensureDbConnection()
+	return source, nil
```

Comment:
This is by design. This is a cyclic dep. You can't create properly initialized InitialResources plugin without having data source, you can't create API Server with InitialResources, you can't create InfluxDB without API Server and you can't create data source without InfluxDB.